File: 2020/csharp/Aoc2020/Day01.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Aoc2020
{
    public class Day01 : IDay
    {
        public override string Day { get; } = nameof(Day01)[3..];

        public override string Result1()
        {
            int[] expenseReport = InputLines.Select((line) => int.Parse(line)).ToArray();
            var (entry1, entry2) = FindTwoEntriesWithSum(2020, expenseReport);
            return (entry1 * entry2).ToString();
        }

        public override string Result2()
        {
            int[] expenseReport = InputLines.Select((line) => int.Parse(line)).ToArray();
            var (entry1, entry2, entry3) = FindThreeEntriesWithSum(2020, expenseReport);
            return (entry1 * entry2 * entry3).ToString();
        }

        static public (int, int) FindTwoEntriesWithSum(int sum, int[] entries)
        {
            return Utils.Combine2(entries).Where(sumIs2020).First();

            bool sumIs2020((int, int) entry) => entry.Item1 + entry.Item2 == sum;
        }

        static public (int, int, int) FindThreeEntriesWithSum(int sum, int[] entries)
        {
            return Combine3(entries).Where(sumIs2020).First();

            bool sumIs2020((int, int, int) entry) => entry.Item1 + entry.Item2 + entry.Item3 == sum;
        }

        static public IEnumerable<(int, int, int)> Combine3(IEnumerable<int> entries)
        {
            var (head, tail) = (entries.First(), entries.Skip(1));
            while (tail.Any())
            {
                foreach (var (first, second) in Utils.Combine2(tail))
                {
                    yield return (head, first, second);
                }
                (head, tail) = (tail.First(), tail.Skip(1));
            }
        }
    }
}

File: 2020/fsharp/Aoc2020/Day01.fs
module Day01

open Utils

let pairWithItem items first =
    seq {
        if items |> (not << Seq.isEmpty) then
            for second in items do
                yield first, second
    }

let pairWithTuple tuples first =
    seq {
        if tuples |> (not << Seq.isEmpty) then
            for second, third in tuples do
                yield first, second, third
    }

let combine items' =
    let items = items' |> Seq.toArray

    items
    |> Seq.indexed
    |> Seq.collect (fun (index, first) ->
        first |> pairWithItem (items |> Seq.skip (index + 1)))

let combine2 items' =
    let items = items' |> Seq.toArray

    items
    |> Seq.indexed
    |> Seq.collect (fun (index, first) ->
        first |> pairWithTuple (items |> Seq.skip (index + 1) |> combine))

let job1 () =
    readInputLines "01"
    |> Seq.map int
    |> combine
    |> Seq.filter (fun (a, b) -> a + b = 2020)
    |> Seq.head
    |> (fun (a, b) -> a * b)
    |> string

let job2 () =
    readInputLines "01"
    |> Seq.map int
    |> combine2
    |> Seq.filter (fun (a, b, c) -> a + b + c = 2020)
    |> Seq.head
    |> (fun (a, b, c) -> a * b * c)
    |> string
